inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,339 @@
+      az_ulib_callback_context flush_callback_context);
+
+  /** Concrete `read` implementation. */
+  az_result (*read)(
+      az_ulib_ustream_forward* ustream_forward,
+      uint8_t* const buffer,
+      size_t buffer_length,
+      size_t* const size);
+
+  /** Concrete `get_size` implementation. */
+  az_result (*get_size)(
+        az_ulib_ustream_forward* ustream_forward, 
+        size_t* const size);
+
+  /** Concrete `dispose` implementation. */
+  az_result (*dispose)(
+        az_ulib_ustream_forward* ustream_forward);
+} az_ulib_ustream_forward_interface;
+
+/**
+ * @brief   Pointer to the data from which to read
+ *
+ * void pointer to memory where the data is located or any needed controls to access the data.
+ * The content of the memory to which this points is up to the ustream_forward implementation.
+ *
+ */
+typedef void* az_ulib_ustream_forward_data;
+
+/**
+ * @brief   Structure for data control block
+ *
+ * For any given ustream_forward that is created, one control block is created and initialized.
+ *
+ * @note    This structure should be viewed and used as internal to the implementation of the
+ *          ustream_forward. Users should therefore not act on it directly and only allocate the 
+ *          memory necessary for it to be passed to the ustream_forward.
+ *
+ */
+struct az_ulib_ustream_forward_tag
+{
+  struct 
+  {
+    /** The #az_ulib_ustream_forward_interface* for this ustream_forward instance type. */
+    const az_ulib_ustream_forward_interface* api;
+
+    /** The #az_ulib_ustream_forward_data* pointing to the data to read. It can be anything that a 
+     * give ustream_forward implementation needs to access the data, whether it be a memory address 
+     * to a buffer, another struct with more controls, etc */
+    const az_ulib_ustream_forward_data* ptr;
+
+    /** The #az_ulib_release_callback to call to release `ptr` once the `ref_count` goes to 
+     * zero. */
+    az_ulib_release_callback data_release;
+
+    /** The #az_ulib_release_callback to call to release the #az_ulib_ustream_forward_data_cb once 
+     * the `ref_count` goes to zero */
+    az_ulib_release_callback ustream_forward_release;
+
+    /** The #offset_t used to keep track of the current position (next returned position). */
+    offset_t inner_current_position;
+
+    /** The `size_t` with the length of the data in the control_block. */
+    size_t length;
+  } _internal;
+};
+
+/**
+ * @brief   Check if a handle is the same type of the API.
+ *
+ *  It will return true if the handle is valid and it is the same type of the API. It will
+ *      return false if the handle is `NULL` or not the correct type.
+ */
+#define AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(handle, type_api)                                     \
+  (!((handle == NULL) || (handle->_internal.api == NULL) \
+     || (handle->_internal.api != &type_api)))
+
+/**
+ * @brief Copy data directly from source (provider) to destination (consumer)
```

Comment:
>Copy data directly from source (provider) to destination (consumer) [](http://example.com/codeflow?start=10&length=67)

I don't think `flush` copy the data, it just forward the data to the flush_callback in one or multiple chunks. Copy the data is one of the things that the flush_callback can do with the received data.